cmd/argoexec/commands/emissary.go

```diff
@@ -209,32 +209,96 @@ func NewEmissaryCommand() *cobra.Command {
 	}
 }
 
-func createCommand(name string, args []string, template *wfv1.Template) (*exec.Cmd, *os.File, *os.File, error) {
+func startCommand(name string, args []string, template *wfv1.Template) (*exec.Cmd, func(), error) {
+	tty := term.IsTerminal(int(os.Stdin.Fd()))
+
 	command := exec.Command(name, args...)
 	command.Env = os.Environ()
 	command.SysProcAttr = &syscall.SysProcAttr{}
-	osspecific.Setpgid(command.SysProcAttr)
-	command.Stdout = os.Stdout
-	command.Stderr = os.Stderr
+	if !tty {
+		// avoid the error "Inappropriate ioctl for device" when
+		// running in tty
+		//
+		// pty.Start uses setsid internally, which makes the process
+		// the group leader already
+		osspecific.Setpgid(command.SysProcAttr)
+	}
+
+	var closer func() = func() {}
+	var stdout io.Writer = os.Stdout
+	var stderr io.Writer = os.Stderr
 
-	var stdout *os.File
-	var combined *os.File
-	var err error
 	// this may not be that important an optimisation, except for very long logs we don't want to capture
 	if includeScriptOutput || template.SaveLogsAsArtifact() {
 		logger.Info("capturing logs")
-		stdout, err = os.OpenFile(varRunArgo+"/ctr/"+containerName+"/stdout", os.O_CREATE|os.O_WRONLY|os.O_APPEND, 0666)
+		stdoutf, err := os.OpenFile(varRunArgo+"/ctr/"+containerName+"/stdout", os.O_CREATE|os.O_WRONLY|os.O_APPEND, 0666)
 		if err != nil {
-			return nil, nil, nil, fmt.Errorf("failed to open stdout: %w", err)
+			return nil, nil, fmt.Errorf("failed to open stdout: %w", err)
 		}
-		combined, err = os.OpenFile(varRunArgo+"/ctr/"+containerName+"/combined", os.O_CREATE|os.O_WRONLY|os.O_APPEND, 0666)
+		combinedf, err := os.OpenFile(varRunArgo+"/ctr/"+containerName+"/combined", os.O_CREATE|os.O_WRONLY|os.O_APPEND, 0666)
 		if err != nil {
-			return nil, nil, nil, fmt.Errorf("failed to open combined: %w", err)
+			return nil, nil, fmt.Errorf("failed to open combined: %w", err)
+		}
+		stdout = io.MultiWriter(stdout, stdoutf, combinedf)
+		stderr = io.MultiWriter(stderr, combinedf)
+
+		closer = func() {
+			_ = stdoutf.Close()
+			_ = combinedf.Close()
 		}
-		command.Stdout = io.MultiWriter(os.Stdout, stdout, combined)
-		command.Stderr = io.MultiWriter(os.Stderr, combined)
 	}
-	return command, stdout, combined, nil
+
+	if tty {
+		logger.Info("container requested TTY")
+		ptmx, err := pty.Start(command)
+		if err != nil {
+			return nil, nil, err
+		}
+
+		// Handle pty size
+		sigWinchCh := make(chan os.Signal, 1)
+		signal.Notify(sigWinchCh, syscall.SIGWINCH)
```

Comment:
does this work on Windows?

Reply:
Do you mean with `kubectl attach` running from a windows host? or with a windows container? I've not tested either. 

AFAICT, this should work fine with `kubectl attach` from a windows host, because `kubectl` has a facility for capturing resizes and sending them to the k8s api in a platform agnostic way. 

As far as running inside a windows container, even if I were familiar with how it differed from linux I have no way of testing it.